drivers/sensor/lsm6dsr/lsm6dsr.c

```diff
@@ -0,0 +1,710 @@
+		return -EIO;
+	}
+
+	if (data->hw_tf->update_reg(dev,
+				    LSM6DSR_REG_CTRL3_C,
+				    LSM6DSR_MASK_CTRL3_C_BDU |
+				    LSM6DSR_MASK_CTRL3_C_IF_INC,
+				    (1 << LSM6DSR_SHIFT_CTRL3_C_BDU) |
+				    (1 << LSM6DSR_SHIFT_CTRL3_C_IF_INC)) < 0) {
+		LOG_DBG("failed to set BDU and burst");
+		return -EIO;
+	}
+
+	if (data->hw_tf->update_reg(dev,
+				    LSM6DSR_REG_CTRL6_C,
+				    LSM6DSR_MASK_CTRL6_C_XL_HM_MODE,
+				    (1 << LSM6DSR_SHIFT_CTRL6_C_XL_HM_MODE)) < 0) {
+		LOG_DBG("failed to disable accelerometer high performance mode");
+		return -EIO;
+	}
+
+	if (data->hw_tf->update_reg(dev,
+				    LSM6DSR_REG_CTRL7_G,
+				    LSM6DSR_MASK_CTRL7_G_HM_MODE,
+				    (1 << LSM6DSR_SHIFT_CTRL7_G_HM_MODE)) < 0) {
+		LOG_DBG("failed to disable gyroscope high performance mode");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int lsm6dsr_init(const struct device *dev)
+{
+	int ret;
+	const struct lsm6dsr_config * const config = dev->config;
+
+	ret = config->bus_init(dev);
+	if (ret < 0) {
+		LOG_ERR("Failed to initialize sensor bus");
+		return ret;
+	}
+
+	ret = lsm6dsr_init_chip(dev);
+	if (ret < 0) {
+		LOG_ERR("Failed to initialize chip");
+		return ret;
+	}
+
+#ifdef CONFIG_LSM6DSR_TRIGGER
+	ret = lsm6dsr_init_interrupt(dev);
+	if (ret < 0) {
+		LOG_ERR("Failed to initialize interrupt.");
+		return ret;
+	}
+#endif
+
+#ifdef CONFIG_LSM6DSR_SENSORHUB
+	ret = lsm6dsr_shub_init_external_chip(dev);
+	if (ret < 0) {
+		LOG_ERR("Failed to initialize external chip");
+		return ret;
+	}
+#endif
+
+	return 0;
+}
+
+/*
+ * Device creation macro, shared by LSM6DSR_DEFINE_SPI() and
+ * LSM6DSR_DEFINE_I2C().
+ */
+
+#define LSM6DSR_DEVICE_INIT(inst)					\
+	SENSOR_DEVICE_DT_INST_DEFINE(inst,				\
+			    lsm6dsr_init,				\
+			    NULL,					\
+			    &lsm6dsr_data_##inst,			\
+			    &lsm6dsr_config_##inst,			\
+			    POST_KERNEL,				\
+			    CONFIG_SENSOR_INIT_PRIORITY,		\
+			    &lsm6dsr_driver_api);
+
+#if DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 0
+#warning "LSM6DSR driver enabled without any devices"
+#endif
+
+
+#ifdef CONFIG_LSM6DSR_TRIGGER
+#define LSM6DSR_CFG_IRQ(inst) \
+		.int_gpio = GPIO_DT_SPEC_INST_GET(inst, irq_gpios),
+#else
+#define LSM6DSR_CFG_IRQ(inst)
+#endif /* CONFIG_LSM6DSR_TRIGGER */
+
+
+/*
+ * Instantiation macros used when a device is on an I2C bus.
+ */
+
+#define LSM6DSR_CONFIG_I2C(inst)					\
+	{								\
+		.bus_init = lsm6dsr_i2c_init,				\
+		.bus_cfg.i2c = I2C_DT_SPEC_INST_GET(inst),		\
+		COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, irq_gpios),	\
+		(LSM6DSR_CFG_IRQ(inst)), ())				\
+	}
+
+#define LSM6DSR_DEFINE_I2C(inst)					\
+	static struct lsm6dsr_data lsm6dsr_data_##inst;			\
+	static const struct lsm6dsr_config lsm6dsr_config_##inst =	\
+		LSM6DSR_CONFIG_I2C(inst);				\
+	LSM6DSR_DEVICE_INIT(inst)
+
+/*
+ * Main instantiation macro. Use of COND_CODE_1() selects the right
+ * bus-specific macro at preprocessor time.
+ */
+
+#define LSM6DSR_DEFINE(inst)	LSM6DSR_DEFINE_I2C(inst)
```

Comment:
Add also SPI support, pls. And once that you create DT properties in dts/bindings add here the loading. You can refer to lsm6dso or others ST driver examples. Moreover in this file you need to use STMEMSC HAL APIs